tools/ds-identify

```diff
@@ -777,21 +777,24 @@ check_config() {
     if [ "$1" = "$files" -a ! -f "$1" ]; then
         return 1
     fi
-    local fname="" line="" ret="" found=0 found_fn=""
-    # shellcheck disable=2094
-    for fname in "$@"; do
-        [ -f "$fname" ] || continue
-        while read line; do
-            line=${line%%#*}
-            case "$line" in
-                $key:\ *|"${key}":)
-                    ret=${line#*:};
-                    ret=${ret# };
-                    found=$((found+1))
-                    found_fn="$fname";;
-            esac
-        done <"$fname"
+    local line="" ret="" found=0 found_fn="" oifs="$IFS" out=""
+    out=$(grep "$key\"\?:" "$@" 2>/dev/null)
```

Comment:
we are ignoring the exit code of grep here. grep will exit 0 on "found", 1 on "not found", 2 on file-not-found. Unfortunately, nothing is checking if the file globs provided to this function exist.  Previously the for loop would stat each expanded glob member. 

When shell expands a glob, any non-matching globs remain in the expanded form.  so if you called with
 
    /path/to/existing.cfg, /path/to/not-existing-glob*, /path/to/not-existing-file

The result "$@" will have

    /path/to/existing.cfg /path/to/not-existing-glob* /path/to/not-existing-file

We're just passing that list right through to grep without checking it.  Checking/filtering it would cost another fork.

bash has an option that is useful for this ('nullglob') but that's not posix /bin/sh.

Reply:
If you're interested, this change checks grep exit code and logs an error:

```diff
diff --git a/tools/ds-identify b/tools/ds-identify
index ad028766f..dcf350a8e 100755
--- a/tools/ds-identify
+++ b/tools/ds-identify
@@ -765,20 +765,35 @@ check_config() {
     # file_globs are expanded via path expansion and
     # default to /etc/cloud/cloud.cfg /etc/cloud/cloud.cfg.d/*.cfg
     # currently does not respect any hierarchy in searching for key.
-    local key="$1" files=""
+    local key="$1" files="" f=""
     shift
     if [ $# -eq 0 ]; then
         files="${PATH_ETC_CI_CFG} ${PATH_ETC_CI_CFG_D}/*.cfg"
     else
         files="$*"
     fi
-    # shellcheck disable=2086
-    { set +f; set -- $files; set -f; }
-    if [ "$1" = "$files" -a ! -f "$1" ]; then
+
+    # expand the globs and check if files exist.
+    set --
+    set +f
+    for f in $files; do
+       [ -f "$f" ] || continue
+       set -- "$@" "$f"
+    done
+    set -f
+
+    if [ $# -eq 0 ]; then
         return 1
     fi
-    local line="" ret="" found=0 found_fn="" oifs="$IFS" out=""
+    local line="" ret="" found=0 found_fn="" oifs="$IFS" out="" rc=""
     out=$(grep "$key\"\?:" "$@" 2>/dev/null)
+    rc=$?
+    if [ $rc -eq 1 ]; then
+        return 1
+    elif [ $rc -ne 0 ]; then
+        error "Unexpected exit code $rc from grep in check_config"
+    fi
+
     IFS=${CR}
     for line in $out; do
         # drop '# comment'
```